### PR TITLE
Support for XY color and attribute discovery

### PR DESCRIPTION
**Pull request by puzzle-star**
Signed-off-by: Pedro Garcia <pg@puzzle-star.com> (github: puzzle-star)

### CODE REVIEW

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java [hunk 1]
```diff
@@ -88,45 +146,99 @@ public void initializeConverter() {
...
+
+                        if(level != currentHSB.getBrightness()) {
+                            changeBrightness(level);
+                            // Wait for transition to complete
+                            // Some lights do not like receiving a level/color change command
+                            // while the previous transition is in progress...
+                            Thread.sleep(1100);
+                        }
 
-                    clusterLevelControl.moveToLevelWithOnOffCommand((int) (level * 254.0 / 100.0 + 0.5), 10).get();
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:126]: Why did you change this? Doesn't it work with some bulbs? I would have preferred to reduce the number of dependant clusters being used unless it's necessary to also use the OnOff?

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java [hunk 2]
```diff
@@ -88,45 +146,99 @@ public void initializeConverter() {
...
+        }
+
         clusterLevelControl.getCurrentLevel(0);
+        clusterOnOff.getOnOff(0);
     }
 
     @Override
     public Runnable handleCommand(final Command command) {
         return new Runnable() {
+            protected void changeOnOff(OnOffType onoff) throws InterruptedException, ExecutionException {
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:170]: Why not make all these methods in the main class rather than this runnable? I plan to change the way this works (I've actually already done so) so I would prefer them as just private methods in the converter class (why are they just protected by the way?).

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java [hunk 3]
```diff
@@ -88,45 +146,99 @@ public void initializeConverter() {
...
+                if(onoff == OnOffType.ON) clusterOnOff.onCommand();
+                else clusterOnOff.offCommand();
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:172]: The formatter needs to be run on this I think? I think the formatter would add {...} around each clause and use separate lines - please use the formatter if it's not used now.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java [hunk 4]
```diff
@@ -135,43 +247,167 @@ public void run() {
...
+                logger.warn("{}: Cannot determine whether device endpoint supports RGB color. Assuming it does by now (checking again later)", endpoint.getEndpointAddress());
+            }
+            else if(clusterColorControl.getSupportedAttributes().contains(ZclColorControlCluster.ATTR_CURRENTHUE)) {
+                logger.debug("{}: Device endpoint supports HUE/SATURATION color", endpoint.getEndpointAddress());
+            }
+            else if(clusterColorControl.getSupportedAttributes().contains(ZclColorControlCluster.ATTR_CURRENTX)) {
+                logger.debug("{}: Device endpoint supports XY color", endpoint.getEndpointAddress());
+            }
+            else {
+                logger.debug("{}: Device endpoint does not support RGB color (probably only color temperature)", endpoint.getEndpointAddress());
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:268]: Does this really need this log entry? I don't think we log a "channel not supported" type message for all other converters.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java [hunk 5]
```diff
@@ -135,43 +247,167 @@ public void run() {
...
+        }
+
         return createChannel(thingUID, endpoint, ZigBeeBindingConstants.CHANNEL_COLOR_COLOR,
                 ZigBeeBindingConstants.ITEM_TYPE_COLOR, "Color");
     }
 
+    protected void updateOnOff(OnOffType onoff) {
+        HSBType oldHSB = currentHSB;
+        updateChannelState(onoff);
+        logger.debug("{}: OnOff updated. {} -> {}", endpoint.getEndpointAddress(), oldHSB.toString(), onoff.toString());
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:283]: I'd suggest to remove this log and add one in either the updateChannelState method, or the setChannelState method so there's a consistent log entry.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java [hunk 6]
```diff
@@ -135,43 +247,167 @@ public void run() {
...
+    }
+
+    protected void updateBrightness(PercentType brightness) {
+        HSBType oldHSB = currentHSB;
+        HSBType newHSB = new HSBType(oldHSB.getHue(), oldHSB.getSaturation(), brightness);
+        currentHSB = newHSB;
+        updateChannelState(brightness);
+        updateChannelState(newHSB);
+        logger.debug("{}: Brightness updated. {} -> {}", endpoint.getEndpointAddress(), oldHSB.toString(), newHSB.toString());
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:292]: I'd suggest to remove this log and add one in either the updateChannelState method, or the setChannelState method so there's a consistent log entry.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java [hunk 7]
```diff
@@ -135,43 +247,167 @@ public void run() {
...
+    }
+
+    protected void updateColorHSB(DecimalType hue, PercentType saturation) {
+        HSBType oldHSB = currentHSB;
+        HSBType newHSB = new HSBType(hue, saturation, oldHSB.getBrightness());
+        currentHSB = newHSB;
+        updateChannelState(newHSB);
+        logger.debug("{}: Color updated. {} -> {}", endpoint.getEndpointAddress(), oldHSB.toString(), newHSB.toString());
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:300]: I'd suggest to remove this log and add one in either the updateChannelState method, or the setChannelState method so there's a consistent log entry.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java [hunk 8]
```diff
@@ -135,43 +247,167 @@ public void run() {
...
+    }
+
+    protected void updateColorXY(PercentType x, PercentType y) {
+        logger.debug("{}: Update Color XY. {}, {}", endpoint.getEndpointAddress(), x.toString(), y.toString());
+        // float hsb[] = ColorHelper.xy2hsv(lastX, lastY);
+        float rgb[] = ColorHelper.xy2rgb(x.floatValue() / 100.0f, y.floatValue() / 100.0f);
+        float hsb[] = ColorHelper.rgb2hsv(rgb[0], rgb[1], rgb[2]);
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:307]: There's already a method to do this in the HSBType in ESH. I wonder if we shouldn't look to add new methods to this class rather than add a separate ColorHelper class to ZigBee (ie add the XY methods to the existing class)?

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorTemperature.java [hunk 9]
```diff
@@ -97,9 +99,24 @@ public void run() {
...
+        if (clusterColorControl == null) {
             return null;
         }
+
+        try {
+            if(!clusterColorControl.discoverAttributes(false).get()) {
+                logger.warn("{}: Cannot determine whether device endpoint supports color temperature. Assuming it does", endpoint.getEndpointAddress());
+            }
+            else if(!clusterColorControl.getSupportedAttributes().contains(ZclColorControlCluster.ATTR_COLORTEMPERATURE)) {
+                logger.debug("{}: Device endpoint does not support color temperature (probably only RGB color)", endpoint.getEndpointAddress());
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorTemperature.java:112]: I'm not sure we really need this log entry as we dont normally log channels that aren't supported?

File: src/main/java/org/openhab/binding/zigbee/converter/color/ColorHelper.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
+/**
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/color/ColorHelper.java:1]: I wonder if this shouldn't be merged into the current HSBType class? This already supports some of the conversions you're adding (ie HSB <> RGB) so if we add the XY conversions into this class, to me that makes more sense. If not, we should at least try and reuse the existing conversions that are available.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java [hunk 11]
```diff
@@ -135,43 +247,167 @@ public void run() {
...
+        clusterColorControl = (ZclColorControlCluster) endpoint.getInputCluster(ZclColorControlCluster.CLUSTER_ID);
+
+        if (clusterColorControl == null || endpoint.getInputCluster(ZclOnOffCluster.CLUSTER_ID) == null
                 || endpoint.getInputCluster(ZclLevelControlCluster.CLUSTER_ID) == null) {
             return null;
         }
+
+        try {
+            if(!clusterColorControl.discoverAttributes(false).get()) {
+                logger.warn("{}: Cannot determine whether device endpoint supports RGB color. Assuming it does by now (checking again later)", endpoint.getEndpointAddress());
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:259]: This probably should be a debug level.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java [hunk 12]
```diff
@@ -61,23 +76,66 @@ public void initializeConverter() {
...
 
+        clusterOnOff = (ZclOnOffCluster) endpoint.getInputCluster(ZclOnOffCluster.CLUSTER_ID);
+        if (clusterOnOff == null) {
+            logger.error("Error opening device on/off controls {}", endpoint.getIeeeAddress());
+            return;
+        }
+
+        try {
+            if(!clusterColorControl.discoverAttributes(false).get()) {
+                logger.warn("{}: Cannot determine whether device endpoint supports RGB color. Assuming it supports HUE/SAT", endpoint.getEndpointAddress());
```
Comment (cdjackson) [inline src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:87]: This probably should be a debug level.

**Review by cdjackson — CHANGES_REQUESTED** [review]
Thanks for this. I've had an initial look - I'd like to spend some more time reviewing this when I get home to see how this works before merging as there's a lot of change in here.

This PR should be against the main openhab repo - not my private repo. I'd suggest to make any updates from this initial review and then submit a new PR into the OH repo.

I think the ColorHelper class would be better added to ESH as part of the HSBType class. There are already methods in this class to do some of the conversions you have added, so we're duplicating code which isn't nice. I think it might be an idea to add the XY methods to this class?

I suspect that the Eclipse formatter hasn't been run on this? If not, please can you run it.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 12 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:87]: Here it is the discovery itself failing, so we are assuming the endpoint supports Hue/Sat when it may not. In my understanding discovery should not normally fail, so hence the "warn", but I can put it as debug.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 2 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:170]: No particular reason. These are called only from the Runnable, so I had put them there, but I will move them to the main class. Will make them private also.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 3 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:172]: Not using eclipse, just plain commandline javac, jar, and vi as editor. Did not think I would need to install the full dev platform for this, but looks like I will have to :)

Comment (cdjackson) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:172]: Ouch - vi? Really ;)

I would really recommend a 'proper' IDE as it will help report various coding errors as well as things like formatting...

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 1 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:126]: It is still there, in the "changeBrightness" method. I had to change a bit the flowr (and add a delay if both the level and color are changed) as at least the IKEA lights do not like receiving a color update while still changing color (and the other way round). It is on the previous commentes in the code

                        if(level != currentHSB.getBrightness()) {
                            changeBrightness(level);
                            // Wait for transition to complete
                            // Some lights do not like receiving a level/color change command
                            // while the previous transition is in progress...
                            Thread.sleep(1100);
                        }

                        if(supportsHue) changeColorHueSaturation(color);
                        else changeColorXY(color);

A different question is why adding the onoff cluster. The reason is to be able to listen to onoff commands and update the channel accordingly (i.e. I may turn the light on just with the IKEA controller, and the binding would not realise).

But I can make this optional, so that if the cluster is not there, ignore it instead of failing.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 11 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:259]: Same as above. Here it is the discovery itself failing, so we are assuming the endpoint supports RGB color when it may not. In my understanding discovery should not normally fail, so hence the "warn", but I can put it as debug.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 4 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:268]: Useful for me to debug while coding, but I can remove it for the PR

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 5 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:283]: Will do

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 6 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:292]: Will do

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 7 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:300]: Will do

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 8 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:307]: Should then be in the HSBType. Will move it there.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorTemperature.java (hunk 9 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorTemperature.java:112]: Useful for me to debug while coding, but I can remove it for the PR

File: src/main/java/org/openhab/binding/zigbee/converter/color/ColorHelper.java (hunk 10 above)
Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/color/ColorHelper.java:1]: As discussed above, will merge into the HSBType class.

File: src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java (hunk 1 above)
Comment (cdjackson) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:126]: I thought I read in the ZCL docs that if a device supports both OnOff and LEvelControl, that they should be linked - so if you send an OnOff command to turn off the device, then I thought that LevelControl report would also be issued? Are you saying that this is not the case in the Ikea bulbs at least or have you not checked this?

If the Ikea bulbs don't work this way, then I'd be happy to leave the OnOff cluster in the converter, and just ignore this somehow if the OnOff isn't supported (I'm not sure how often OnOff won't be supported if LevelControl is, but it just seems best to reduce dependancies if possible)

Comment (puzzle-star) [reply src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java:126]: As far as I remember OnOff and LevelControl are linked, but sending an On/Off command does not need to report a change in Level, as the level does get changed necessarily by default when an On/Off command is sent (last level is used when On). At least the Ikea lights do not send a level attribute update here, and either my flspp led RGB controller.

So I think we will have to keep onoff cluster here

**Comment by puzzle-star** [comment]
I will install eclipse to work on these changes: it will take a couple of days anyway, as eveything has a learning curve, and besides I have a work to do during the daytime :) It's a pity, because I love using vi :)

Apart from this, I have a question on the behaviour of the ColorTemperature channel. It is designed as a dimmer, I understand for convinience, lacking a better item type. But it becomes sort of a nuisance when using a colortemp dimmable bulb, as we can only keep one state in the item (also, what does on/off mean then?).

By doing so, we need then to configure both channels (ColorTemp and SwitchLevel) and putting them in a group (unlike the ColorChannel, which manages on/off, level, and color). Also, due to this grouping aggregation fails (the group, i.e., a Group:Switch:OR(ON:OFF) item, thinks it is On due to colortemp not being zero.

Would it make more sense for it to be Color instead of Dimmer, and use either HSBType (which can contain a color temperature conversion) or a DecimalType for color Temperature in mireds ? Will not change it by now, but do you think it may be worth opening an issue to see if there is some work ongoing to support ColorTemp as a channel itself ?

**Comment by cdjackson** [comment]
You like Vi ;) Wow ;) I know - it’s just whatever you get used to…

Maybe discuss the color temp stuff on ESH forum, or issue list. I agree it’s not real nice at the moment - I was also thinking that a CT to HSB conversion in the HSBType might also be useful so we can display CT on the color ‘wheel’ (it’s not perfect either, but it would give some level of ‘compatibility'). I would suggest to look at what the Hue binding does in ESH repo as that’s probably what we’ll end up having to use.

Types in ESH are a bit difficult (IMHO!) and there’s often a reluctance to add more (I can understand this, but the problem is the current types often don’t suit the use case very well at all).

Thanks

**Comment by cdjackson** [comment]
Closing as this is now discussed in the OH repo
